### ISSUE TITLE
Shallow water regression tests often fail on GPU

### ISSUE THREAD

**Issue by tomchor** [CliMA/Oceananigans.jl#2922]
This isn't a huge issue but I've noticed recently that often the shallow water regression tests on GPU fail with the following error:

```
Shallow Water Bickley jet simulation [GPU, ConservativeFormulation]: Test Failed at /net/ocean/home/data44/data5/glwagner/.buildkite-agent/builds/sverdrup-6/clima/oceananigans/test/regression_tests/shallow_water_bickley_jet_regression.jl:93
  | Expression: all(test_fields.v .≈ truth_fields.v)
  | Stacktrace:
  | [1] macro expansion
  | @ /net/ocean/home/data44/data5/glwagner/julia-1.8.5/share/julia/stdlib/v1.8/Test/src/Test.jl:464 [inlined]
  | [2] run_shallow_water_regression(arch::GPU, formulation::ConservativeFormulation; regenerate_data::Bool)
  | @ Main /net/ocean/home/data44/data5/glwagner/.buildkite-agent/builds/sverdrup-6/clima/oceananigans/test/regression_tests/shallow_water_bickley_jet_regression.jl:93
```

Sometimes I need to restart the test twice for it to pass. Like I said, it's not a huge issue, but do we understand why that's happening? Is there anything we can do to prevent that behavior?

**Comment by vchuravy** [comment]
@simone-silvestri and I think there is a missing synchronization for one of the boundary updates

**Comment by simone-silvestri** [comment]
I am also doubting our regression data. For sure it is a synchronization problem, but it is difficult to understand wether it's on the data (generated long time ago) or in the current code

**Comment by simone-silvestri** [comment]
You are also right, the data was generated on the CPU so it should be synchronized correctly.

**Comment by simone-silvestri** [comment]
It might be also another problem not related to BC. 
When the test passes we get
```
[2023/02/12 17:00:57.335] INFO  Δv: min=-2.709402e-11, max=+2.834735e-11, mean=-2.606411e-14, absmean=+1.007853e-12, std=+2.676095e-12 (16512/16512 matching grid points)
```
On test failing
```
[2023/02/14 10:02:14.671] INFO  Δv: min=-3.212770e-11, max=+2.834735e-11, mean=-2.751160e-13, absmean=+1.256905e-12, std=+3.884381e-12 (16384/16512 matching grid points)
```
(same maximum difference larger mean). I will check which points are actually not matching, it might help us

**Comment by simone-silvestri** [comment]
running on my new PR (#2888) I get this for the v velocity

`diff = Bool.(1 .- (truth_fields.v .≈ test_fields.v))`

![Screenshot 2023-02-14 at 10 41 48 AM](https://user-images.githubusercontent.com/33547697/218787450-f6426596-03bc-4680-bc6b-89e9be925357.png)


which seems to suggets that it is not a BC issue

**Comment by navidcy** [comment]
just for v? for u is good?

**Comment by simone-silvestri** [comment]
This is the same quantity evaluated in PR #2926 

![Uploading Screenshot 2023-02-14 at 11.09.54 AM.png…]()

```
[2023/02/14 11:07:05.226] INFO  Δv: min=-2.345130e-05, max=+4.749146e-05, mean=+1.747027e-06, absmean=+3.169314e-06, std=+8.512165e-06 (10620/16512 matching grid points)
```

Yep, u is good and also h. The weird thing is that
```
[2023/02/14 11:06:15.550] INFO  Δu: min=-1.671992e-04, max=+4.237079e-06, mean=-8.685825e-06, absmean=+8.778487e-06, std=+2.661356e-05 (16384/16384 matching grid points)
[2023/02/14 11:06:15.550] INFO  Δv: min=-2.347391e-05, max=+4.751164e-05, mean=+1.746998e-06, absmean=+3.169677e-06, std=+8.513593e-06 (10752/16512 matching grid points)
[2023/02/14 11:06:15.550] INFO  Δh: min=-1.340684e-04, max=+1.454704e-04, mean=+2.980232e-08, absmean=+1.029445e-05, std=+2.835074e-05 (16384/16384 matching grid points)
```
So u and h match but with a larger difference

**Comment by simone-silvestri** [comment]
maximum iterations for this test are 20 with a 128 grid size, so I think we need to exclude the boundary condition possibility and look at other synchronization issues (maybe particular to the shallow water model)

**Comment by glwagner** [comment]
Since we don't have bitwise reproducibility, each run has random round-off errors that, if integrated into a nonlinear trajectory, can cause two solution to diverge eventually. Possibly this test case is a little too challenging (too long time integration, or too nonlinear), meaning that occasional chaotic divergence of the solution is expected?

**Comment by glwagner** [comment]
There's a tradeoff between taking too few time-steps (not enough time for subtle differences to manifest above the noise floor) and taking too many time steps. Perhaps just halving the time-steps for this test will make it more robust but hopefully still useful

**Comment by simone-silvestri** [comment]
I agree, let's just disable this test for now so we can merge outstanding PRs and rewrite the test

**Comment by navidcy** [comment]
Do you reckon the problems stem from #2890?

**Comment by navidcy** [comment]
On Australian GPUs

```

julia> run_shallow_water_regression(arch, ConservativeFormulation(); regenerate_data = false)
[2023/02/15 08:09:30.057] INFO  Initializing simulation...
[2023/02/15 08:09:30.137] INFO      ... simulation initialization complete (79.862 ms)
[2023/02/15 08:09:30.137] INFO  Executing initial time step...
[2023/02/15 08:09:49.128] INFO      ... initial time step complete (18.991 seconds).
[2023/02/15 08:09:49.218] INFO  Simulation is stopping after running for 19.159 seconds.
[2023/02/15 08:09:49.218] INFO  Model iteration 20 equals or exceeds stop iteration 20.
[2023/02/15 08:09:49.338] INFO  Δu: min=-2.979433e-08, max=+2.978377e-08, mean=-1.025389e-10, absmean=+1.911258e-09, std=+5.570608e-09 (16384/16384 matching grid points)
[2023/02/15 08:09:49.338] INFO  Δv: min=-2.709283e-11, max=+2.834728e-11, mean=-2.606411e-14, absmean=+1.007855e-12, std=+2.676083e-12 (16512/16512 matching grid points)
[2023/02/15 08:09:49.339] INFO  Δh: min=-4.765317e-07, max=+4.756881e-07, mean=+1.373701e-08, absmean=+2.044855e-07, std=+2.513230e-07 (16384/16384 matching grid points)
Test Passed

julia> run_shallow_water_regression(arch, formulation; regenerate_data = false)VectorInvariantFormulation()^C

julia> run_shallow_water_regression(arch, VectorInvariantFormulation(); regenerate_data = false)
[2023/02/15 08:10:46.432] WARN  Inflating model grid halo size to (4, 4, 0) and recreating grid. Note that an ImmersedBoundaryGrid requires an extra halo point.  -@-> /g/data/v45/nc3020/Oceananigans.jl/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:233
[2023/02/15 08:10:46.449] INFO  Initializing simulation...
[2023/02/15 08:10:46.450] INFO      ... simulation initialization complete (580.963 μs)
[2023/02/15 08:10:46.450] INFO  Executing initial time step...
[2023/02/15 08:10:46.453] INFO      ... initial time step complete (3.038 ms).
[2023/02/15 08:10:46.490] INFO  Simulation is stopping after running for 39.012 ms.
[2023/02/15 08:10:46.490] INFO  Model iteration 20 equals or exceeds stop iteration 20.
[2023/02/15 08:10:46.492] INFO  Δu: min=-2.910407e-08, max=+2.965560e-08, mean=+4.423683e-11, absmean=+1.880294e-09, std=+5.514070e-09 (16384/16384 matching grid points)
[2023/02/15 08:10:46.492] INFO  Δv: min=-2.905665e-11, max=+2.481851e-11, mean=-8.868305e-14, absmean=+1.029733e-12, std=+2.802455e-12 (16512/16512 matching grid points)
[2023/02/15 08:10:46.492] INFO  Δh: min=-4.765574e-07, max=+4.768079e-07, mean=+3.492460e-09, absmean=+2.039438e-07, std=+2.510525e-07 (16384/16384 matching grid points)
Test Passed
```

Weird.

Why we get that warning for the VectorInvariant formulation?

**Comment by simone-silvestri** [comment]
because vector invariant advects vorticity that requires one extra point for vorticity calculation (we should remove that inflate and just throw an error). 

On what branch is this?

**Comment by simone-silvestri** [comment]
> Do you reckon the problems stem from https://github.com/CliMA/Oceananigans.jl/pull/2890?

Oh well, that could be... but probably it would error on CPUs as well? I can test it out

**Comment by navidcy** [comment]
> On what branch is this?

`main`

**Comment by tomchor** [comment]
> On Australian GPUs
> 
> ```julia
> julia> run_shallow_water_regression(arch, ConservativeFormulation(); regenerate_data = false)
> [2023/02/15 08:09:30.057] INFO  Initializing simulation...
> [2023/02/15 08:09:30.137] INFO      ... simulation initialization complete (79.862 ms)
> [2023/02/15 08:09:30.137] INFO  Executing initial time step...
> [2023/02/15 08:09:49.128] INFO      ... initial time step complete (18.991 seconds).
> [2023/02/15 08:09:49.218] INFO  Simulation is stopping after running for 19.159 seconds.
> [2023/02/15 08:09:49.218] INFO  Model iteration 20 equals or exceeds stop iteration 20.
> [2023/02/15 08:09:49.338] INFO  Δu: min=-2.979433e-08, max=+2.978377e-08, mean=-1.025389e-10, absmean=+1.911258e-09, std=+5.570608e-09 (16384/16384 matching grid points)
> [2023/02/15 08:09:49.338] INFO  Δv: min=-2.709283e-11, max=+2.834728e-11, mean=-2.606411e-14, absmean=+1.007855e-12, std=+2.676083e-12 (16512/16512 matching grid points)
> [2023/02/15 08:09:49.339] INFO  Δh: min=-4.765317e-07, max=+4.756881e-07, mean=+1.373701e-08, absmean=+2.044855e-07, std=+2.513230e-07 (16384/16384 matching grid points)
> Test Passed
> 
> julia> run_shallow_water_regression(arch, formulation; regenerate_data = false)VectorInvariantFormulation()^C
> 
> julia> run_shallow_water_regression(arch, VectorInvariantFormulation(); regenerate_data = false)
> [2023/02/15 08:10:46.432] WARN  Inflating model grid halo size to (4, 4, 0) and recreating grid. Note that an ImmersedBoundaryGrid requires an extra halo point.  -@-> /g/data/v45/nc3020/Oceananigans.jl/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:233
> [2023/02/15 08:10:46.449] INFO  Initializing simulation...
> [2023/02/15 08:10:46.450] INFO      ... simulation initialization complete (580.963 μs)
> [2023/02/15 08:10:46.450] INFO  Executing initial time step...
> [2023/02/15 08:10:46.453] INFO      ... initial time step complete (3.038 ms).
> [2023/02/15 08:10:46.490] INFO  Simulation is stopping after running for 39.012 ms.
> [2023/02/15 08:10:46.490] INFO  Model iteration 20 equals or exceeds stop iteration 20.
> [2023/02/15 08:10:46.492] INFO  Δu: min=-2.910407e-08, max=+2.965560e-08, mean=+4.423683e-11, absmean=+1.880294e-09, std=+5.514070e-09 (16384/16384 matching grid points)
> [2023/02/15 08:10:46.492] INFO  Δv: min=-2.905665e-11, max=+2.481851e-11, mean=-8.868305e-14, absmean=+1.029733e-12, std=+2.802455e-12 (16512/16512 matching grid points)
> [2023/02/15 08:10:46.492] INFO  Δh: min=-4.765574e-07, max=+4.768079e-07, mean=+3.492460e-09, absmean=+2.039438e-07, std=+2.510525e-07 (16384/16384 matching grid points)
> Test Passed
> ```
> 
> Weird.
> 
> Why we get that warning for the VectorInvariant formulation?

@navidcy do you know the GPU and driver info for these tests?

As a reference [here's the info for tests on buildkite](https://github.com/CliMA/Oceananigans.jl/pull/2900#issuecomment-1418288880):

```
glwagner@sverdrup:~$ nvidia-smi -q

==============NVSMI LOG==============

Timestamp                                 : Sun Feb  5 17:58:59 2023
Driver Version                            : 465.27
CUDA Version                              : 11.3

Attached GPUs                             : 1
GPU 00000000:82:00.0
    Product Name                          : NVIDIA Quadro P6000
    Product Brand                         : Quadro
```

**Comment by simone-silvestri** [comment]
did you use the test environment or Oceananigans's enviroment? On Tartarus, branch main in the test environment does this

```
Precompiling project...
  1 dependency successfully precompiled in 38 seconds. 218 already precompiled.
     Testing Running tests...
[NVBLAS] NVBLAS_CONFIG_FILE environment variable is NOT set : relying on default config filename 'nvblas.conf'
[NVBLAS] Cannot open default config file 'nvblas.conf'
[NVBLAS] Config parsed
[NVBLAS] CPU Blas library need to be provided
[2023/02/14 16:22:39.804] INFO  Running shallow water regression tests...
[2023/02/14 16:22:40.097] INFO    Testing shallow water Bickley jet simulation regression [GPU, VectorInvariantFormulation]
[2023/02/14 16:23:08.854] WARN  Inflating model grid halo size to (4, 4, 0) and recreating grid. Note that an ImmersedBoundaryGrid requires an extra halo point.  -@-> /home/ssilvest/stable_oceananigans/Oceananigans.jl/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:233
[2023/02/14 16:24:01.792] INFO  Initializing simulation...
[2023/02/14 16:24:08.481] INFO      ... simulation initialization complete (6.689 seconds)
[2023/02/14 16:24:08.481] INFO  Executing initial time step...
[2023/02/14 16:24:43.227] INFO      ... initial time step complete (34.746 seconds).
[2023/02/14 16:24:43.387] INFO  Simulation is stopping after running for 41.593 seconds.
[2023/02/14 16:24:43.387] INFO  Model iteration 20 equals or exceeds stop iteration 20.
[2023/02/14 16:24:48.950] INFO  Δu: min=-1.671992e-04, max=+4.237079e-06, mean=-8.685825e-06, absmean=+8.778487e-06, std=+2.661356e-05 (16384/16384 matching grid points)
[2023/02/14 16:24:48.951] INFO  Δv: min=-2.347391e-05, max=+4.751164e-05, mean=+1.746998e-06, absmean=+3.169677e-06, std=+8.513593e-06 (10752/16512 matching grid points)
[2023/02/14 16:24:48.952] INFO  Δh: min=-1.340684e-04, max=+1.454704e-04, mean=+2.980232e-08, absmean=+1.029445e-05, std=+2.835074e-05 (16384/16384 matching grid points)
Shallow Water Bickley jet simulation [GPU, VectorInvariantFormulation]: Test Failed at /home/ssilvest/stable_oceananigans/Oceananigans.jl/test/regression_tests/shallow_water_bickley_jet_regression.jl:93
  Expression: all(test_fields.v .≈ truth_fields.v)
Stacktrace:
 [1] macro expansion
   @ ~/julia-1.8.0/share/julia/stdlib/v1.8/Test/src/Test.jl:464 [inlined]
 [2] run_shallow_water_regression(arch::GPU, formulation::VectorInvariantFormulation; regenerate_data::Bool)
   @ Main ~/stable_oceananigans/Oceananigans.jl/test/regression_tests/shallow_water_bickley_jet_regression.jl:93
[2023/02/14 16:24:51.806] INFO    Testing shallow water Bickley jet simulation regression [GPU, ConservativeFormulation]
[2023/02/14 16:25:16.858] INFO  Initializing simulation...
[2023/02/14 16:25:16.992] INFO      ... simulation initialization complete (134.028 ms)
[2023/02/14 16:25:16.992] INFO  Executing initial time step...
[2023/02/14 16:25:47.962] INFO      ... initial time step complete (30.969 seconds).
[2023/02/14 16:25:48.168] INFO  Simulation is stopping after running for 31.304 seconds.
[2023/02/14 16:25:48.168] INFO  Model iteration 20 equals or exceeds stop iteration 20.
[2023/02/14 16:25:48.190] INFO  Δu: min=-1.672000e-04, max=+4.242261e-06, mean=-8.684623e-06, absmean=+8.777260e-06, std=+2.661194e-05 (16384/16384 matching grid points)
[2023/02/14 16:25:48.191] INFO  Δv: min=-2.345113e-05, max=+4.749113e-05, mean=+1.747024e-06, absmean=+3.169309e-06, std=+8.512156e-06 (10620/16512 matching grid points)
[2023/02/14 16:25:48.191] INFO  Δh: min=-1.338214e-04, max=+1.449336e-04, mean=-4.033232e-17, absmean=+1.028773e-05, std=+2.832882e-05 (16384/16384 matching grid points)
Shallow Water Bickley jet simulation [GPU, ConservativeFormulation]: Test Failed at /home/ssilvest/stable_oceananigans/Oceananigans.jl/test/regression_tests/shallow_water_bickley_jet_regression.jl:93
  Expression: all(test_fields.v .≈ truth_fields.v)
Stacktrace:
 [1] macro expansion
   @ ~/julia-1.8.0/share/julia/stdlib/v1.8/Test/src/Test.jl:464 [inlined]
 [2] run_shallow_water_regression(arch::GPU, formulation::ConservativeFormulation; regenerate_data::Bool)
   @ Main ~/stable_oceananigans/Oceananigans.jl/test/regression_tests/shallow_water_bickley_jet_regression.jl:93
Test Summary:                                                                | Pass  Fail  Total     Time
Oceananigans                                                                 |    4     2      6  3m14.4s
  Single file test                                                           |    4     2      6  3m14.4s
    Shallow Water Regression                                                 |    4     2      6  3m09.1s
      Shallow Water Bickley jet simulation [GPU, VectorInvariantFormulation] |    2     1      3  2m11.7s
      Shallow Water Bickley jet simulation [GPU, ConservativeFormulation]    |    2     1      3    56.4s
```

**Comment by navidcy** [comment]
Interesting. No I didn't use the test env. I run in the repl...

**Comment by simone-silvestri** [comment]
What surprises me is that not only `v` is different. Despite passing, also the summary for `u` and `h` is very different.
actually also outside the test environment, I get the same results 😅 .  

I guess this confirms that this test is a bit too hard and we just should decrease the number of time steps

**Comment by simone-silvestri** [comment]
I could try to regenerate the data with the version of Oceananigans we used the last time (I guess it's this one https://github.com/CliMA/Oceananigans.jl/releases/tag/v0.76.5) but only 10 time steps and see if the results change

**Comment by navidcy** [comment]
```
nc3020@gadi-gpu-v100-0104:~$ nvidia-smi -q

==============NVSMI LOG==============

Timestamp                                 : Wed Feb 15 08:59:13 2023
Driver Version                            : 525.60.13
CUDA Version                              : 12.0

Attached GPUs                             : 1
GPU 00000000:B2:00.0
    Product Name                          : Tesla V100-SXM2-32GB
    Product Brand                         : Tesla
    Product Architecture                  : Volta
    Display Mode                          : Enabled
    Display Active                        : Disabled
    Persistence Mode                      : Enabled
...
```

**Comment by navidcy** [comment]
Tests fail on v0.76.5 on GPU... weird...

```Julia
nc3020@gadi-gpu-v100-0100:/g/data/v45/nc3020/OC.jl$ julia-1.6 --project
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.6.7 (2022-07-19)
 _/ |\__'_|_|_|\__'_|  |
|__/                   |

(Oceananigans) pkg> test
     Testing Oceananigans
      Status `/jobfs/73413669.gadi-pbs/jl_XhyZPR/Project.toml`
  [79e6a3ab] Adapt v3.3.3
  [6e4b80f9] BenchmarkTools v1.3.2
  [052768ef] CUDA v3.10.0
  [72cfdca4] CUDAKernels v0.3.3
  [a2441757] Coverage v1.6.0
  [a8cc5b0e] Crayons v4.1.1
  [7445602f] CubedSphere v0.2.0
  [124859b0] DataDeps v0.7.10
  [ffbed154] DocStringExtensions v0.8.6
  [7a1cc6ca] FFTW v1.4.6
  [c27321d9] Glob v1.3.0
  [40713840] IncompleteLU v0.2.0
  [42fd0dbc] IterativeSolvers v0.9.2
  [033835bb] JLD2 v0.4.22
  [63c18a36] KernelAbstractions v0.7.2
  [da04e1cc] MPI v0.19.2
  [85f8d34a] NCDatasets v0.12.4
  [9e8cae18] Oceananigans v0.76.5 `/g/data/v45/nc3020/OC.jl`
  [6fe1bfb0] OffsetArrays v1.11.2
  [bac558e1] OrderedCollections v1.4.1
  [0e08944d] PencilArrays v0.17.2
  [4a48f351] PencilFFTs v0.13.6
  [91a5bcdd] Plots v1.36.0
  [6038ab10] Rotations v1.3.1
  [1bc83da4] SafeTestsets v0.0.1
  [d496a93d] SeawaterPolynomials v0.2.3
  [09ab397b] StructArrays v0.6.7
  [a759f4b9] TimerOutputs v0.5.19
  [bdfc003b] TimesDates v0.3.1
  [bc48ee85] Tullio v0.3.4
  [ade2ca70] Dates `@stdlib/Dates`
  [b77e0a4c] InteractiveUtils `@stdlib/InteractiveUtils`
  [37e2e46d] LinearAlgebra `@stdlib/LinearAlgebra`
  [56ddb016] Logging `@stdlib/Logging`
  [44cfe95a] Pkg `@stdlib/Pkg`
  [de0858da] Printf `@stdlib/Printf`
  [9a3f8284] Random `@stdlib/Random`
  [2f01184e] SparseArrays `@stdlib/SparseArrays`
  [10745b16] Statistics `@stdlib/Statistics`
  [8dfed614] Test `@stdlib/Test`
      Status `/jobfs/73413669.gadi-pbs/jl_XhyZPR/Manifest.toml`
  [621f4979] AbstractFFTs v1.1.0
  [79e6a3ab] Adapt v3.3.3
  [4fba245c] ArrayInterface v6.0.1
  [30b0a656] ArrayInterfaceCore v0.1.2
  [ab4f0b2a] BFloat16s v0.2.0
  [6e4b80f9] BenchmarkTools v1.3.2
  [d1d4a3ce] BitFlags v0.1.7
  [fa961155] CEnum v0.4.2
  [179af706] CFTime v0.1.2
  [052768ef] CUDA v3.10.0
  [72cfdca4] CUDAKernels v0.3.3
  [49dc2e85] Calculus v0.5.1
  [7057c7e9] Cassette v0.3.10
  [d360d2e6] ChainRulesCore v1.15.0
  [9e997f8a] ChangesOfVariables v0.1.3
  [944b1d66] CodecZlib v0.7.1
  [35d6a980] ColorSchemes v3.20.0
  [3da002f7] ColorTypes v0.11.4
  [c3611d14] ColorVectorSpace v0.9.10
  [5ae59095] Colors v0.12.10
  [34da2185] Compat v3.43.0
  [a216cea6] CompoundPeriods v0.5.1
  [d38c429a] Contour v0.6.2
  [a2441757] Coverage v1.6.0
  [c36e975a] CoverageTools v1.3.0
  [a8cc5b0e] Crayons v4.1.1
  [7445602f] CubedSphere v0.2.0
  [9a962f9c] DataAPI v1.10.0
  [124859b0] DataDeps v0.7.10
  [864edb3b] DataStructures v0.18.12
  [e2d170a0] DataValueInterfaces v1.0.0
  [b552c78f] DiffRules v1.11.0
  [ffbed154] DocStringExtensions v0.8.6
  [fa6b7ba4] DualNumbers v0.6.8
  [b305315f] Elliptic v1.0.1
  [e2ba6199] ExprTools v0.1.8
  [c87230d0] FFMPEG v0.4.1
  [7a1cc6ca] FFTW v1.4.6
  [5789e2e9] FileIO v1.14.0
  [53c48c17] FixedPointNumbers v0.8.4
  [59287772] Formatting v0.4.2
  [0c68f7d7] GPUArrays v8.3.2
  [61eb1bfa] GPUCompiler v0.15.2
  [28b8d3ca] GR v0.69.5
  [c27321d9] Glob v1.3.0
  [42e2da0e] Grisu v1.0.2
  [cd3eb016] HTTP v1.7.4
  [615f187c] IfElse v0.1.1
  [40713840] IncompleteLU v0.2.0
  [83e8ac13] IniFile v0.5.1
  [842dd82b] InlineStrings v1.4.0
  [3587e190] InverseFunctions v0.1.4
  [92d709cd] IrrationalConstants v0.1.1
  [42fd0dbc] IterativeSolvers v0.9.2
  [82899510] IteratorInterfaceExtensions v1.0.0
  [033835bb] JLD2 v0.4.22
  [1019f520] JLFzf v0.1.5
  [692b3bcd] JLLWrappers v1.4.1
  [682c06a0] JSON v0.21.3
  [0f8b85d8] JSON3 v1.9.5
  [63c18a36] KernelAbstractions v0.7.2
  [929cbde3] LLVM v4.11.1
  [b964fa9f] LaTeXStrings v1.3.0
  [23fbe1c1] Latexify v0.15.18
  [2ab3a3ac] LogExpFunctions v0.3.15
  [e6f89c97] LoggingExtras v1.0.0
  [da04e1cc] MPI v0.19.2
  [3da0fdf6] MPIPreferences v0.1.7
  [1914dd2f] MacroTools v0.5.9
  [739be429] MbedTLS v1.1.7
  [442fdcdd] Measures v0.3.2
  [e1d29d7a] Missings v1.1.0
  [78c3b35d] Mocking v0.7.5
  [85f8d34a] NCDatasets v0.12.4
  [77ba4419] NaNMath v1.0.0
  [9e8cae18] Oceananigans v0.76.5 `/g/data/v45/nc3020/OC.jl`
  [6fe1bfb0] OffsetArrays v1.11.2
  [4d8831e6] OpenSSL v1.3.3
  [bac558e1] OrderedCollections v1.4.1
  [69de0a69] Parsers v2.3.1
  [0e08944d] PencilArrays v0.17.2
  [4a48f351] PencilFFTs v0.13.6
  [b98c9c47] Pipe v1.3.0
  [ccf2f8ad] PlotThemes v3.1.0
  [995b91a9] PlotUtils v1.3.4
  [91a5bcdd] Plots v1.36.0
  [21216c6a] Preferences v1.3.0
  [94ee1d12] Quaternions v0.5.6
  [74087812] Random123 v1.5.0
  [e6cf234a] RandomNumbers v1.5.3
  [3cdcf5f2] RecipesBase v1.2.1
  [01d81517] RecipesPipeline v0.6.9
  [189a3867] Reexport v1.2.2
  [05181044] RelocatableFolders v1.0.0
  [ae029012] Requires v1.3.0
  [6038ab10] Rotations v1.3.1
  [1bc83da4] SafeTestsets v0.0.1
  [6c6a2e73] Scratch v1.1.1
  [d496a93d] SeawaterPolynomials v0.2.3
  [992d4aef] Showoff v1.0.3
  [777ac1f9] SimpleBufferStream v1.1.0
  [66db9d55] SnoopPrecompile v1.0.3
  [a2af1166] SortingAlgorithms v1.1.0
  [276daf66] SpecialFunctions v2.1.5
  [aedffcd0] Static v0.6.5
  [90137ffa] StaticArrays v1.4.4
  [15972242] StaticPermutations v0.3.0
  [82ae8749] StatsAPI v1.5.0
  [2913bbd2] StatsBase v0.33.21
  [5e0ebb24] Strided v1.2.2
  [09ab397b] StructArrays v0.6.7
  [856f2bd8] StructTypes v1.8.1
  [3783bdb8] TableTraits v1.0.1
  [bd369af6] Tables v1.7.0
  [6aa5eb33] TaylorSeries v0.10.13
  [62fd8b95] TensorCore v0.1.1
  [f269a46b] TimeZones v1.9.1
  [a759f4b9] TimerOutputs v0.5.19
  [bdfc003b] TimesDates v0.3.1
  [3bb67fe8] TranscodingStreams v0.9.6
  [bc48ee85] Tullio v0.3.4
  [9d95972d] TupleTools v1.3.0
  [5c2747f8] URIs v1.4.2
  [1cfade01] UnicodeFun v0.4.1
  [41fe7b60] Unzip v0.1.2
  [81def892] VersionParsing v1.3.0
  [6e34b625] Bzip2_jll v1.0.8+0
  [83423d85] Cairo_jll v1.16.1+1
  [2e619515] Expat_jll v2.4.8+0
  [b22a6f82] FFMPEG_jll v4.4.2+2
  [f5851436] FFTW_jll v3.3.10+0
  [a3f928ae] Fontconfig_jll v2.13.93+0
  [d7e528f0] FreeType2_jll v2.10.4+0
  [559328eb] FriBidi_jll v1.0.10+0
  [0656b61e] GLFW_jll v3.3.8+0
  [d2c73de3] GR_jll v0.69.1+0
  [78b55507] Gettext_jll v0.21.0+0
  [7746bdde] Glib_jll v2.74.0+2
  [3b182d85] Graphite2_jll v1.3.14+0
  [0234f1f7] HDF5_jll v1.12.0+1
  [2e76f6c2] HarfBuzz_jll v2.8.1+1
  [1d5cc7b8] IntelOpenMP_jll v2018.0.3+2
  [aacddb02] JpegTurbo_jll v2.1.91+0
  [c1c5ebd0] LAME_jll v3.100.1+0
  [88015f11] LERC_jll v3.0.0+1
  [dad2f222] LLVMExtra_jll v0.0.16+2
  [dd4b983a] LZO_jll v2.10.1+0
  [e9f186c6] Libffi_jll v3.2.2+1
  [d4300ac3] Libgcrypt_jll v1.8.7+0
  [7e76a0d4] Libglvnd_jll v1.6.0+0
  [7add5ba3] Libgpg_error_jll v1.42.0+0
  [94ce4f54] Libiconv_jll v1.16.1+2
  [4b2f31a3] Libmount_jll v2.35.0+0
  [89763e89] Libtiff_jll v4.4.0+0
  [38a345b3] Libuuid_jll v2.36.0+0
  [856f044c] MKL_jll v2022.0.0+0
  [7cb0a576] MPICH_jll v4.0.2+5
  [9237b28f] MicrosoftMPI_jll v10.1.3+2
  [7243133f] NetCDF_jll v400.702.400+0
  [e7412a2a] Ogg_jll v1.3.5+1
  [fe0851c0] OpenMPI_jll v4.1.3+3
  [458c3c95] OpenSSL_jll v1.1.14+0
  [efe28fd5] OpenSpecFun_jll v0.5.5+0
  [91d4177d] Opus_jll v1.3.2+0
  [30392449] Pixman_jll v0.40.1+0
  [ea2cea3b] Qt5Base_jll v5.15.3+2
  [a2964d1f] Wayland_jll v1.21.0+0
  [2381bf8a] Wayland_protocols_jll v1.25.0+0
  [02c8fc9c] XML2_jll v2.10.3+0
  [aed1982a] XSLT_jll v1.1.34+0
  [4f6342f7] Xorg_libX11_jll v1.6.9+4
  [0c0b7dd1] Xorg_libXau_jll v1.0.9+4
  [935fb764] Xorg_libXcursor_jll v1.2.0+4
  [a3789734] Xorg_libXdmcp_jll v1.1.3+4
  [1082639a] Xorg_libXext_jll v1.3.4+4
  [d091e8ba] Xorg_libXfixes_jll v5.0.3+4
  [a51aa0fd] Xorg_libXi_jll v1.7.10+4
  [d1454406] Xorg_libXinerama_jll v1.1.4+4
  [ec84b674] Xorg_libXrandr_jll v1.5.2+4
  [ea2f1a96] Xorg_libXrender_jll v0.9.10+4
  [14d82f49] Xorg_libpthread_stubs_jll v0.1.0+3
  [c7cfdc94] Xorg_libxcb_jll v1.13.0+3
  [cc61e674] Xorg_libxkbfile_jll v1.1.0+4
  [12413925] Xorg_xcb_util_image_jll v0.4.0+1
  [2def613f] Xorg_xcb_util_jll v0.4.0+1
  [975044d2] Xorg_xcb_util_keysyms_jll v0.4.0+1
  [0d47668e] Xorg_xcb_util_renderutil_jll v0.3.9+1
  [c22f9ab0] Xorg_xcb_util_wm_jll v0.4.1+1
  [35661453] Xorg_xkbcomp_jll v1.4.2+4
  [33bec58e] Xorg_xkeyboard_config_jll v2.27.0+4
  [c5fb5394] Xorg_xtrans_jll v1.4.0+3
  [3161d3a3] Zstd_jll v1.5.4+0
  [214eeab7] fzf_jll v0.29.0+0
  [a4ae2306] libaom_jll v3.4.0+0
  [0ac62f75] libass_jll v0.15.1+0
  [f638f0a6] libfdk_aac_jll v2.0.2+0
  [b53b4c65] libpng_jll v1.6.38+0
  [f27f6e37] libvorbis_jll v1.3.7+1
  [1270edf5] x264_jll v2021.5.5+0
  [dfaa095f] x265_jll v3.5.0+0
  [d8fb68d0] xkbcommon_jll v1.4.1+0
  [0dad84c5] ArgTools `@stdlib/ArgTools`
  [56f22d72] Artifacts `@stdlib/Artifacts`
  [2a0f44e3] Base64 `@stdlib/Base64`
  [ade2ca70] Dates `@stdlib/Dates`
  [8bb1440f] DelimitedFiles `@stdlib/DelimitedFiles`
  [8ba89e20] Distributed `@stdlib/Distributed`
  [f43a241f] Downloads `@stdlib/Downloads`
  [b77e0a4c] InteractiveUtils `@stdlib/InteractiveUtils`
  [4af54fe1] LazyArtifacts `@stdlib/LazyArtifacts`
  [b27032c2] LibCURL `@stdlib/LibCURL`
  [76f85450] LibGit2 `@stdlib/LibGit2`
  [8f399da3] Libdl `@stdlib/Libdl`
  [37e2e46d] LinearAlgebra `@stdlib/LinearAlgebra`
  [56ddb016] Logging `@stdlib/Logging`
  [d6f4376e] Markdown `@stdlib/Markdown`
  [a63ad114] Mmap `@stdlib/Mmap`
  [ca575930] NetworkOptions `@stdlib/NetworkOptions`
  [44cfe95a] Pkg `@stdlib/Pkg`
  [de0858da] Printf `@stdlib/Printf`
  [9abbd945] Profile `@stdlib/Profile`
  [3fa0cd96] REPL `@stdlib/REPL`
  [9a3f8284] Random `@stdlib/Random`
  [ea8e919c] SHA `@stdlib/SHA`
  [9e88b42a] Serialization `@stdlib/Serialization`
  [1a1011a3] SharedArrays `@stdlib/SharedArrays`
  [6462fe0b] Sockets `@stdlib/Sockets`
  [2f01184e] SparseArrays `@stdlib/SparseArrays`
  [10745b16] Statistics `@stdlib/Statistics`
  [4607b0f0] SuiteSparse `@stdlib/SuiteSparse`
  [fa267f1f] TOML `@stdlib/TOML`
  [a4e569a6] Tar `@stdlib/Tar`
  [8dfed614] Test `@stdlib/Test`
  [cf7118a7] UUIDs `@stdlib/UUIDs`
  [4ec0a83e] Unicode `@stdlib/Unicode`
  [e66e0078] CompilerSupportLibraries_jll `@stdlib/CompilerSupportLibraries_jll`
  [deac9b47] LibCURL_jll `@stdlib/LibCURL_jll`
  [29816b5a] LibSSH2_jll `@stdlib/LibSSH2_jll`
  [c8ffd9c3] MbedTLS_jll `@stdlib/MbedTLS_jll`
  [14a3606d] MozillaCACerts_jll `@stdlib/MozillaCACerts_jll`
  [05823500] OpenLibm_jll `@stdlib/OpenLibm_jll`
  [efcefdf7] PCRE2_jll `@stdlib/PCRE2_jll`
  [83775a58] Zlib_jll `@stdlib/Zlib_jll`
  [8e850ede] nghttp2_jll `@stdlib/nghttp2_jll`
  [3f19e933] p7zip_jll `@stdlib/p7zip_jll`
Precompiling project...
  ✗ RecipesPipeline
  ✗ Plots
  0 dependencies successfully precompiled in 7 seconds (205 already precompiled)
  2 dependencies errored. To see a full report either run `import Pkg; Pkg.precompile()` or load the packages
     Testing Running tests...
┌ Warning: You appear to be using MPI.jl with the default MPI binary on a cluster.
│ We recommend using the system-provided MPI, see the Configuration section of the MPI.jl docs.
└ @ MPI /g/data/v45/nc3020/.julia/packages/MPI/08SPr/deps/deps.jl:15
[ Info: Oceananigans will use 48 threads
[2023/02/21 06:36:13.652] WARN  allowscalar([true]) is deprecated, use `allowscalar() do end` or `@allowscalar` to denote exactly which operations can use scalar operations. -@-> /g/data/v45/nc3020/OC.jl/test/dependencies_for_runtests.jl:71
[2023/02/21 06:36:18.673] WARN  allowscalar([true]) is deprecated, use `allowscalar() do end` or `@allowscalar` to denote exactly which operations can use scalar operations. -@-> /g/data/v45/nc3020/OC.jl/test/dependencies_for_runtests.jl:71
[2023/02/21 06:36:18.728] WARN  Over-writing registration of the datadep -@-> /g/data/v45/nc3020/.julia/packages/DataDeps/ae6dT/src/registration.jl:15
[2023/02/21 06:36:18.729] WARN  Over-writing registration of the datadep -@-> /g/data/v45/nc3020/.julia/packages/DataDeps/ae6dT/src/registration.jl:15
[2023/02/21 06:36:19.136] INFO  Running shallow water regression tests...
[2023/02/21 06:36:19.143] INFO    Testing shallow water Bickley jet simulation regression [GPU, VectorInvariantFormulation]
[2023/02/21 06:36:29.566] WARN  defaulting to uniform WENO scheme with Float64 precision, use WENO5(grid = grid) if this was not intended -@-> /g/data/v45/nc3020/OC.jl/src/Advection/weno_fifth_order.jl:187
[2023/02/21 06:36:32.345] WARN  defaulting to uniform WENO scheme with Float64 precision, use WENO5(grid = grid) if this was not intended -@-> /g/data/v45/nc3020/OC.jl/src/Advection/weno_fifth_order.jl:187
[2023/02/21 06:36:32.346] WARN  defaulting to uniform WENO scheme with Float64 precision, use WENO5(grid = grid) if this was not intended -@-> /g/data/v45/nc3020/OC.jl/src/Advection/weno_fifth_order.jl:187
[2023/02/21 06:37:27.833] INFO  Initializing simulation...
[2023/02/21 06:37:32.539] INFO      ... simulation initialization complete (4.632 seconds)
[2023/02/21 06:37:32.539] INFO  Executing initial time step...
[2023/02/21 06:37:52.296] WARN  `query(s::CuStream)` is deprecated, use `isdone(s)` instead. -@-> /g/data/v45/nc3020/.julia/packages/CUDAKernels/kCOA4/src/CUDAKernels.jl:41
[2023/02/21 06:37:54.934] INFO      ... initial time step complete (22.395 seconds).
[2023/02/21 06:37:56.880] INFO  Simulation is stopping. Model iteration 20 has hit or exceeded simulation stop iteration 20.
[2023/02/21 06:38:00.743] INFO  Δu: min=-3.188983e-08, max=+2.966291e-08, mean=+3.932481e-11, absmean=+1.978398e-09, std=+5.640766e-09 (16384/16384 matching grid points)
[2023/02/21 06:38:00.744] INFO  Δv: min=-2.539346e-09, max=+1.967717e-09, mean=+2.462280e-12, absmean=+1.480658e-10, std=+4.143014e-10 (16488/16512 matching grid points)
[2023/02/21 06:38:00.744] INFO  Δh: min=-4.767265e-07, max=+4.779508e-07, mean=+3.492460e-09, absmean=+2.039559e-07, std=+2.510841e-07 (16384/16384 matching grid points)
Shallow Water Bickley jet simulation [GPU, VectorInvariantFormulation]: Test Failed at /g/data/v45/nc3020/OC.jl/test/regression_tests/shallow_water_bickley_jet_regression.jl:94
  Expression: all(test_fields.v .≈ truth_fields.v)
Stacktrace:
 [1] run_shallow_water_regression(arch::GPU, formulation::VectorInvariantFormulation; regenerate_data::Bool)
   @ Main /g/data/v45/nc3020/OC.jl/test/regression_tests/shallow_water_bickley_jet_regression.jl:94
 [2] macro expansion
   @ /g/data/v45/nc3020/OC.jl/test/test_shallow_water_regression.jl:12 [inlined]
 [3] macro expansion
   @ /g/data/v45/nc3020/julia/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [4] macro expansion
   @ /g/data/v45/nc3020/OC.jl/test/test_shallow_water_regression.jl:11 [inlined]
 [5] macro expansion
   @ /g/data/v45/nc3020/julia/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [6] top-level scope
   @ /g/data/v45/nc3020/OC.jl/test/test_shallow_water_regression.jl:6
[2023/02/21 06:38:01.335] INFO    Testing shallow water Bickley jet simulation regression [GPU, ConservativeFormulation]
[2023/02/21 06:38:01.772] WARN  defaulting to uniform WENO scheme with Float64 precision, use WENO5(grid = grid) if this was not intended -@-> /g/data/v45/nc3020/OC.jl/src/Advection/weno_fifth_order.jl:187
[2023/02/21 06:38:01.916] WARN  defaulting to uniform WENO scheme with Float64 precision, use WENO5(grid = grid) if this was not intended -@-> /g/data/v45/nc3020/OC.jl/src/Advection/weno_fifth_order.jl:187
[2023/02/21 06:38:01.916] WARN  defaulting to uniform WENO scheme with Float64 precision, use WENO5(grid = grid) if this was not intended -@-> /g/data/v45/nc3020/OC.jl/src/Advection/weno_fifth_order.jl:187
[2023/02/21 06:38:30.393] INFO  Initializing simulation...
[2023/02/21 06:38:30.504] INFO      ... simulation initialization complete (110.043 ms)
[2023/02/21 06:38:30.504] INFO  Executing initial time step...
[2023/02/21 06:38:40.776] INFO      ... initial time step complete (10.272 seconds).
[2023/02/21 06:38:42.742] INFO  Simulation is stopping. Model iteration 20 has hit or exceeded simulation stop iteration 20.
[2023/02/21 06:38:46.004] INFO  Δu: min=-3.698171e-08, max=+2.977649e-08, mean=-9.793714e-11, absmean=+2.315996e-09, std=+6.206031e-09 (16384/16384 matching grid points)
[2023/02/21 06:38:46.005] INFO  Δv: min=-5.495966e-10, max=+1.063640e-09, mean=-3.441143e-13, absmean=+4.699188e-11, std=+1.347006e-10 (16508/16512 matching grid points)
[2023/02/21 06:38:46.005] INFO  Δh: min=-4.766520e-07, max=+4.757091e-07, mean=+1.373701e-08, absmean=+2.044884e-07, std=+2.513218e-07 (16384/16384 matching grid points)
Shallow Water Bickley jet simulation [GPU, ConservativeFormulation]: Test Failed at /g/data/v45/nc3020/OC.jl/test/regression_tests/shallow_water_bickley_jet_regression.jl:94
  Expression: all(test_fields.v .≈ truth_fields.v)
Stacktrace:
 [1] run_shallow_water_regression(arch::GPU, formulation::ConservativeFormulation; regenerate_data::Bool)
   @ Main /g/data/v45/nc3020/OC.jl/test/regression_tests/shallow_water_bickley_jet_regression.jl:94
 [2] macro expansion
   @ /g/data/v45/nc3020/OC.jl/test/test_shallow_water_regression.jl:12 [inlined]
 [3] macro expansion
   @ /g/data/v45/nc3020/julia/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [4] macro expansion
   @ /g/data/v45/nc3020/OC.jl/test/test_shallow_water_regression.jl:11 [inlined]
 [5] macro expansion
   @ /g/data/v45/nc3020/julia/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [6] top-level scope
   @ /g/data/v45/nc3020/OC.jl/test/test_shallow_water_regression.jl:6
Test Summary:                                                              | Pass  Fail  Total
Oceananigans                                                               |    4     2      6
  Shallow Water Regression                                                 |    4     2      6
    Shallow Water Bickley jet simulation [GPU, VectorInvariantFormulation] |    2     1      3
    Shallow Water Bickley jet simulation [GPU, ConservativeFormulation]    |    2     1      3
ERROR: LoadError: Some tests did not pass: 4 passed, 2 failed, 0 errored, 0 broken.
in expression starting at /g/data/v45/nc3020/OC.jl/test/runtests.jl:3
ERROR: Package Oceananigans errored during testing

julia> versioninfo()^C

(Oceananigans) pkg> st
     Project Oceananigans v0.76.5
      Status `/g/data/v45/nc3020/OC.jl/Project.toml`
  [79e6a3ab] Adapt v3.3.3
  [052768ef] CUDA v3.10.0
  [72cfdca4] CUDAKernels v0.3.3
  [a8cc5b0e] Crayons v4.1.1
  [7445602f] CubedSphere v0.2.0
  [ffbed154] DocStringExtensions v0.8.6
  [7a1cc6ca] FFTW v1.4.6
  [c27321d9] Glob v1.3.0
  [40713840] IncompleteLU v0.2.0
  [42fd0dbc] IterativeSolvers v0.9.2
  [033835bb] JLD2 v0.4.22
  [63c18a36] KernelAbstractions v0.7.2
  [da04e1cc] MPI v0.19.2
  [85f8d34a] NCDatasets v0.12.4
  [6fe1bfb0] OffsetArrays v1.11.2
  [bac558e1] OrderedCollections v1.4.1
  [0e08944d] PencilArrays v0.17.2
  [4a48f351] PencilFFTs v0.13.6
  [6038ab10] Rotations v1.3.1
  [1bc83da4] SafeTestsets v0.0.1
  [d496a93d] SeawaterPolynomials v0.2.3
  [09ab397b] StructArrays v0.6.7
  [bc48ee85] Tullio v0.3.4
  [ade2ca70] Dates
  [b77e0a4c] InteractiveUtils
  [37e2e46d] LinearAlgebra
  [56ddb016] Logging
  [44cfe95a] Pkg
  [de0858da] Printf
  [9a3f8284] Random
  [2f01184e] SparseArrays
  [10745b16] Statistics

julia> versioninfo()
Julia Version 1.6.7
Commit 3b76b25b64* (2022-07-19 15:11 UTC)
Platform Info:
  OS: Linux (x86_64-redhat-linux)
  CPU: Intel(R) Xeon(R) Platinum 8268 CPU @ 2.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, cascadelake)
Environment:
  JULIA_DEPOT_PATH = /g/data/v45/nc3020/.julia:/share/julia/site/
  JULIA_CUDA_USE_BINARYBUILDER = false
  JULIA_LOAD_PATH = @:@v#.#:@stdlib:@site
  JULIA_NUM_THREADS = 48
```

**Comment by navidcy** [comment]
@simone-silvestri do you also get the test passing if you run on v0.76.5?

I'm quite confused... I'm running using Julia v1.6.7

**Comment by francispoulin** [comment]
@navidcy I am happy to try something to compare.  If yes, what exactly would you like me to test?

**Comment by navidcy** [comment]
OK, I tried to make the test pass and it doesn't on v0.76.5-v0.76.8 it it won't pass. On v0.77.0 it passes. This hints that the data were created with v0.77.0.

@simone-silvestri do you have any recollection of this or a way to figure it out?

**Comment by simone-silvestri** [comment]
Hmmm, I think it was created with the first PR that introduces the test and then left untouched. What if we regenerate the data in main. Will it work on previous versions?

**Comment by navidcy** [comment]
Well doesn't generating the data with `main` beats the purpose a bit? What if we have introduced a bug and then we regenerate the data with the bug and then test against that...?

**Comment by simone-silvestri** [comment]
I meant generating it on main and testing it on 0.76.5 or 0.76.8

**Comment by navidcy** [comment]
oh I see...

**Comment by navidcy** [comment]
yes, I don't see how that would be different. But I can try.

**Comment by navidcy** [comment]
https://github.com/CliMA/Oceananigans.jl/compare/v0.76.5...v0.77.0

So from here I see that there was a change from using `1/V` to `1/A` in the advective fluxes. Would that imply a change?

@simone-silvestri can you see any other difference here that would have an effect to the solution?

**Comment by navidcy** [comment]
Please help 🥹

**Comment by simone-silvestri** [comment]
So main agrees with Artifacts (the previously generated data?)? Is this on a CPU, because it seems like the difference arises on the GPU.

**Comment by navidcy** [comment]
It agrees _but_ tests fail on buildkite which is so confusing

**Comment by glwagner** [comment]
Did you test on sverdrup?

**Comment by navidcy** [comment]
OK, after chatting with @simone-silvestri seems like the data was created with v0.77.0.

So I recreated it with half the tilmestep (1e-2 -> 5e-3) and pushed here --> https://github.com/CliMA/OceananigansArtifacts.jl/pull/19

**Comment by glwagner** [comment]
Ok!

sverdrup also uses a somewhat unusual GPU (Quadro P6000) and there can be differences due to that.

**Comment by navidcy** [comment]
unfortunately I am reopening this since we had an instance of failure once more

**Comment by navidcy** [comment]
@francispoulin I'm wondering if something changed at some point since v0.77.0.

If you can do this it would help me. Could you save the tendencies for the first 2 time steps for $u$, $v$, and $h$ from v0.77.0 and from v0.79.4. You can even use Δt=1. If the tendencies from the initial timestep are different then we may need to dig deeper?

**Comment by francispoulin** [comment]
@navidcy : certainly!  I will look at this today and hope to have something for you by tomorrow.

**Comment by francispoulin** [comment]
I am running `julia 1.8.2` with `Oceananigans v0.79.4`, which I am running from my repo that I cloned.

I find that the same test fails on a `CPU` for either formulations.  When I focus on one test case I see that the differences are almost everywhere, `256/16512`.  Note that I am running the test case from REPL.  Please see below for the details.

I found that the data files that we are comparing with are from June 1 2022.  I wanted to try and reproduce them by going into `shallow_water_bickley_jet_regression.jl` and changed `regenerate_data` from `false` to `true`.  This ran but it didn't save the data.  Where can I find the file that generated the `truth` data?

I made some figures to see whether there is a visual difference between the two and the answer is yes.  I am including `vtruth.png`, `vnew.png` and `vdiff.png`,   One observation is the amplitude of the `truth` is significantly larger and noisier.  I wonder whether the noise that was added before was of larger amplitude than what we have here?

![vdiff](https://user-images.githubusercontent.com/8239041/221380109-96ff99c1-d6cf-4ce5-9b40-52c38eaaca11.png)
![vtruth](https://user-images.githubusercontent.com/8239041/221380111-3ee2224e-3ba3-4169-90e2-3f03cd00e78b.png)
![vnew](https://user-images.githubusercontent.com/8239041/221380112-8a34bccf-4485-406b-92e6-74f6a13f479e.png)




```
julia> include("test_shallow_water_regression.jl")
[2023/02/25 15:42:20.222] WARN  Over-writing registration of the datadep -@-> /home/fpoulin/.julia/packages/DataDeps/ae6dT/src/registration.jl:15
[2023/02/25 15:42:20.223] WARN  Over-writing registration of the datadep -@-> /home/fpoulin/.julia/packages/DataDeps/ae6dT/src/registration.jl:15
[2023/02/25 15:42:20.355] INFO  Running shallow water regression tests...
[2023/02/25 15:42:20.355] INFO    Testing shallow water Bickley jet simulation regression [CPU, ConservativeFormulation]
[2023/02/25 15:42:21.128] INFO  Initializing simulation...
[2023/02/25 15:42:21.129] INFO      ... simulation initialization complete (685.283 μs)
[2023/02/25 15:42:21.129] INFO  Executing initial time step...
[2023/02/25 15:42:23.148] INFO      ... initial time step complete (2.019 seconds).
[2023/02/25 15:42:23.672] INFO  Simulation is stopping after running for 2.521 seconds.
[2023/02/25 15:42:23.672] INFO  Model iteration 20 equals or exceeds stop iteration 20.
[2023/02/25 15:42:23.814] INFO  Δu: min=-1.673332e-04, max=+1.094476e-04, mean=+2.507827e-07, absmean=+5.538768e-06, std=+1.574960e-05 (16384/16384 matching grid points)
[2023/02/25 15:42:23.814] INFO  Δv: min=-2.878156e-04, max=+1.230686e-04, mean=+1.406671e-07, absmean=+2.141572e-05, std=+4.550342e-05 (256/16512 matching grid points)
[2023/02/25 15:42:23.814] INFO  Δh: min=-1.126421e-03, max=+1.182485e-03, mean=+9.255018e-09, absmean=+7.901569e-05, std=+2.067916e-04 (16384/16384 matching grid points)
Shallow Water Bickley jet simulation [CPU, ConservativeFormulation]: Test Failed at /home/fpoulin/Software/Oceananigans.jl/test/regression_tests/shallow_water_bickley_jet_regression.jl:95
  Expression: all(test_fields.v .≈ truth_fields.v)
Stacktrace:
 [1] macro expansion
   @ ~/Software/julia-1.8.2/share/julia/stdlib/v1.8/Test/src/Test.jl:464 [inlined]
 [2] run_shallow_water_regression(arch::CPU, formulation::ConservativeFormulation; regenerate_data::Bool)
   @ Main ~/Software/Oceananigans.jl/test/regression_tests/shallow_water_bickley_jet_regression.jl:95
Test Summary:                                                         | Fail  Total  Time
Shallow Water Regression                                              |    1      1  3.5s
  Shallow Water Bickley jet simulation [CPU, ConservativeFormulation] |    1      1  3.5s
ERROR: LoadError: Some tests did not pass: 0 passed, 1 failed, 0 errored, 0 broken.
in expression starting at /home/fpoulin/Software/Oceananigans.jl/test/test_shallow_water_regression.jl:6
```

**Comment by francispoulin** [comment]
These are plots of the initial condtions from the `truth` (reference), `new` (current solution) and `difference` (difference between the two) for the `u` velocities.  The maximum of the difference seems to be about `3e-4`, and the perturbations that we add in are 1e-4.  This makes me wonder whether the `truth` scenario has larger perturbations by a factor of 3, and hence why the perturbations grow much faster in that scenario?

![udiff_ic](https://user-images.githubusercontent.com/8239041/221384657-a895f8dc-5b7d-4dd1-82d2-a9c7db1245d3.png)
![utruth_ic](https://user-images.githubusercontent.com/8239041/221384658-4f2d2927-e4b6-4be1-bcee-d683a26e1264.png)
![unew_ic](https://user-images.githubusercontent.com/8239041/221384659-a32977b1-eed1-4850-b4ca-734a1a9ed6b3.png)

**Comment by francispoulin** [comment]
I'll only upload the difference for the height field, as the fields are very similar.  Here we see that the difference has  a maximum of `1e-7`, which is very small.  So my theory is that the two simulations do not pick exactly the same magnitude of perturbations for the `u` field.

![hdiff_ic](https://user-images.githubusercontent.com/8239041/221384890-89a5f9d9-70a3-4a19-b12d-06b1a8e8e93d.png)

**Comment by francispoulin** [comment]
I tried to follow the terminology in the regression test and sorry it was confusing.  I used `truth` to refer to the data that is read from a file, done [here.](https://github.com/CliMA/Oceananigans.jl/blob/3676a718be1160f3ea70c3cce5dd21c5f06f144a/test/regression_tests/shallow_water_bickley_jet_regression.jl#L82) 

I agree that only `v` fails but that means that `v` has larger differences compared to `u` and `h`.  They all have differences.  

I compared the initial data that we used in this regression test with the initial data read from the regression test.  I saw that we had the 0th and 20th step saved.  If there are differences at the beginning, then they are not solving exactly the same problem.  In both we have that v is set to 0 and u and h are set to the Bickley jet with a random perturbation on `u`.  The randomness will not be the same (unless we use a seed, which we don't, but we could) however the amplitude of the perturbations are different.  This suggests to me that the initial conditions are not the same, and maybe the soruce of why the regression test fails.  This is why I would like to know how the initial data was generated, using what script.  I do suggest we regenerate it as that might solve a lot of the problems we are having with the regression tests.

Do you want me to generate a script that load the inital data and compare it?  You can see the results above.  The fact that h has differences of `1e-7` is due to single precison, which means they are the same.  However, u is different with errors much larger then `1e-7`, so the initial data is different.  Unless there is something that I'm missing here?

I saw that the data files were dated June 1st 2022.  For me it is stored at the following location

`/home/fpoulin/.julia/datadeps/regression_test_data/`

**Comment by navidcy** [comment]
hm... the data in `.julia/datadeps/regression_test_data/` are downloaded once from the CliMA/OceananigansArtifacts.jl repo. So perhaps you are using some outdated data for your "truth"?

But regardless, if there are differences in the initial condition then we shall look no further! That's the issue!

@simone-silvestri, is there a way to enforce DataDeps to download again?

**Comment by simone-silvestri** [comment]
You can delete the folder with the data. It would be in `.julia/datadeps`

**Comment by francispoulin** [comment]
> You can delete the folder with the data. It would be in `.julia/datadeps`

Thanks!  I will give that a try and let you know how it goes on my end.

**Comment by francispoulin** [comment]
I tried it on a different computer where I installed datadeps brand new and the tests ran on the CPU.  I guess I am catching up to the problem.  Unfortunately, my laptop doesn't have a GPU so I need to run it elsewhere to reproduce the problem.  Will work on that today.

**Comment by francispoulin** [comment]
I removed the `.julia` file on my computer with a GPU and then it passed all 4 regression tests.  

Maybe this needs to be done on the other servers to get the test to pass?

I will point out that for the `VectorInvariantFormulation` the absmean and std are identical to four or more decimals for GPU and CPU.

However, for the `ConservativeFormulation`, the values for `v` and `h` are as good but the `u` value for absmean and std off by about 100.  Maybe this is a concern, even though the tests pass?


```
julia> include("test/test_shallow_water_regression.jl")
[2023/02/27 18:57:22.803] WARN  Over-writing registration of the datadep -@-> /home/fpoulin/.julia/packages/DataDeps/ae6dT/src/registration.jl:15
[2023/02/27 18:57:22.803] WARN  Over-writing registration of the datadep -@-> /home/fpoulin/.julia/packages/DataDeps/ae6dT/src/registration.jl:15
[2023/02/27 18:57:23.118] INFO  Running shallow water regression tests...
    FJP: arch and formulation = CPU()VectorInvariantFormulation()
[2023/02/27 18:57:23.119] INFO    Testing shallow water Bickley jet simulation regression [CPU, VectorInvariantFormulation]
[2023/02/27 18:57:23.124] WARN  Inflating model grid halo size to (4, 4, 0) and recreating grid. Note that an ImmersedBoundaryGrid requires an extra halo point.  -@-> /home/fpoulin/Software/Oceananigans.jl/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:223
[2023/02/27 18:57:23.525] INFO  Initializing simulation...
[2023/02/27 18:57:23.526] INFO      ... simulation initialization complete (1.081 ms)
[2023/02/27 18:57:23.526] INFO  Executing initial time step...
[2023/02/27 18:57:24.661] INFO      ... initial time step complete (1.135 seconds).
[2023/02/27 18:57:46.538] INFO  Simulation is stopping after running for 21.864 seconds.
[2023/02/27 18:57:46.538] INFO  Model iteration 20 equals or exceeds stop iteration 20.
[2023/02/27 18:57:46.680] INFO  Δu: min=-2.958580e-08, max=+2.921001e-08, mean=-7.420989e-11, absmean=+1.893230e-09, std=+5.534008e-09 (16384/16384 matching grid points)
[2023/02/27 18:57:46.680] INFO  Δv: min=-1.426954e-11, max=+1.432324e-11, mean=-2.560603e-14, absmean=+6.119071e-13, std=+1.792743e-12 (16512/16512 matching grid points)
[2023/02/27 18:57:46.680] INFO  Δh: min=-4.754092e-07, max=+4.764754e-07, mean=+2.561137e-09, absmean=+2.079621e-07, std=+2.556520e-07 (16384/16384 matching grid points)
    FJP: arch and formulation = CPU()ConservativeFormulation()
[2023/02/27 18:57:46.680] INFO    Testing shallow water Bickley jet simulation regression [CPU, ConservativeFormulation]
[2023/02/27 18:57:49.567] INFO  Initializing simulation...
[2023/02/27 18:57:49.625] INFO      ... simulation initialization complete (57.503 ms)
[2023/02/27 18:57:49.625] INFO  Executing initial time step...
[2023/02/27 18:58:00.006] INFO      ... initial time step complete (10.381 seconds).
[2023/02/27 18:58:00.502] INFO  Simulation is stopping after running for 10.911 seconds.
[2023/02/27 18:58:00.502] INFO  Model iteration 20 equals or exceeds stop iteration 20.
[2023/02/27 18:58:00.509] INFO  Δu: min=-2.972623e-08, max=+2.977460e-08, mean=+3.472408e-11, absmean=+2.084662e-09, std=+5.973908e-09 (16384/16384 matching grid points)
[2023/02/27 18:58:00.509] INFO  Δv: min=-1.453264e-11, max=+1.436524e-11, mean=+4.659518e-15, absmean=+6.267771e-13, std=+1.805375e-12 (16512/16512 matching grid points)
[2023/02/27 18:58:00.509] INFO  Δh: min=-4.766880e-07, max=+4.767828e-07, mean=-9.313224e-10, absmean=+2.070016e-07, std=+2.552765e-07 (16384/16384 matching grid points)
    FJP: arch and formulation = GPU()VectorInvariantFormulation()
[2023/02/27 18:58:00.510] INFO    Testing shallow water Bickley jet simulation regression [GPU, VectorInvariantFormulation]
[2023/02/27 18:58:00.515] WARN  Inflating model grid halo size to (4, 4, 0) and recreating grid. Note that an ImmersedBoundaryGrid requires an extra halo point.  -@-> /home/fpoulin/Software/Oceananigans.jl/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:223
[2023/02/27 18:58:00.613] INFO  Initializing simulation...
[2023/02/27 18:58:00.613] INFO      ... simulation initialization complete (299.591 μs)
[2023/02/27 18:58:00.613] INFO  Executing initial time step...
[2023/02/27 18:58:00.615] INFO      ... initial time step complete (1.524 ms).
[2023/02/27 18:58:00.636] INFO  Simulation is stopping after running for 21.966 ms.
[2023/02/27 18:58:00.636] INFO  Model iteration 20 equals or exceeds stop iteration 20.
[2023/02/27 18:58:00.644] INFO  Δu: min=-2.958580e-08, max=+2.921001e-08, mean=-7.420989e-11, absmean=+1.893230e-09, std=+5.534008e-09 (16384/16384 matching grid points)
[2023/02/27 18:58:00.644] INFO  Δv: min=-1.426930e-11, max=+1.432286e-11, mean=-2.560596e-14, absmean=+6.119088e-13, std=+1.792748e-12 (16512/16512 matching grid points)
[2023/02/27 18:58:00.644] INFO  Δh: min=-4.754092e-07, max=+4.764754e-07, mean=+2.561137e-09, absmean=+2.079621e-07, std=+2.556520e-07 (16384/16384 matching grid points)
    FJP: arch and formulation = GPU()ConservativeFormulation()
[2023/02/27 18:58:00.644] INFO    Testing shallow water Bickley jet simulation regression [GPU, ConservativeFormulation]
[2023/02/27 18:58:00.731] INFO  Initializing simulation...
[2023/02/27 18:58:00.732] INFO      ... simulation initialization complete (546.689 μs)
[2023/02/27 18:58:00.732] INFO  Executing initial time step...
[2023/02/27 18:58:00.734] INFO      ... initial time step complete (2.363 ms).
[2023/02/27 18:58:00.795] INFO  Simulation is stopping after running for 61.833 ms.
[2023/02/27 18:58:00.795] INFO  Model iteration 20 equals or exceeds stop iteration 20.
[2023/02/27 18:58:00.795] INFO  Δu: min=-2.584649e-06, max=+1.597348e-06, mean=+3.118505e-09, absmean=+1.254639e-07, std=+3.731235e-07 (16384/16384 matching grid points)
[2023/02/27 18:58:00.795] INFO  Δv: min=-1.453306e-11, max=+1.436533e-11, mean=+4.659446e-15, absmean=+6.267764e-13, std=+1.805380e-12 (16512/16512 matching grid points)
[2023/02/27 18:58:00.795] INFO  Δh: min=-4.766880e-07, max=+4.767828e-07, mean=-9.313224e-10, absmean=+2.070016e-07, std=+2.552765e-07 (16384/16384 matching grid points)
Test Summary:            | Pass  Total   Time
Shallow Water Regression |    4      4  37.7s
Test.DefaultTestSet("Shallow Water Regression", Any[Test.DefaultTestSet("Shallow Water Bickley jet simulation [CPU, VectorInvariantFormulation]", Any[], 1, false, false, true, 1.677542243119075e9, 1.677542266680591e9), Test.DefaultTestSet("Shallow Water Bickley jet simulation [CPU, ConservativeFormulation]", Any[], 1, false, false, true, 1.677542266680676e9, 1.677542280509984e9), Test.DefaultTestSet("Shallow Water Bickley jet simulation [GPU, VectorInvariantFormulation]", Any[], 1, false, false, true, 1.677542280510052e9, 1.677542280644663e9), Test.DefaultTestSet("Shallow Water Bickley jet simulation [GPU, ConservativeFormulation]", Any[], 1, false, false, true, 1.677542280644765e9, 1.677542280795948e9)], 0, false, false, true, 1.677542243118905e9, 1.677542280795951e9)
```

**Comment by francispoulin** [comment]
@navidcy : did you want to try deleting the `.julia` folder and then retry the tests to see if this fixes the problem for you?

**Comment by simone-silvestri** [comment]
It should, I can try deleting them and checking

**Comment by francispoulin** [comment]
@simone-silvestri : if you could that would be great.  Otherwise, it is very mysterious that it runs on my local computer but not on your machine. I will try it on a server and confirm that it runs for me.

**Comment by francispoulin** [comment]
I can confirm that I tried this on a server and after freshly installing the data it worked fine.  I should say that I didn't run the tests exactly but ran this code below.  Note the comment was just to make it easier for me to parse the output.

```
using Test
include("dependencies_for_runtests.jl")
include("data_dependencies.jl")

include("regression_tests/shallow_water_bickley_jet_regression.jl")

@testset "Shallow Water Regression" begin
    @info "Running shallow water regression tests..."

    for arch in (CPU(), GPU())
        for formulation in (VectorInvariantFormulation(), ConservativeFormulation())
                print("   FJP: arch and formulation = ", arch, formulation, "\n")
                @testset "Shallow Water Bickley jet simulation [$(typeof(arch)), $(typeof(formulation))]" begin
                @info "  Testing shallow water Bickley jet simulation regression [$(typeof(arch)), $(typeof(formulation))]"
                run_shallow_water_regression(arch, formulation; regenerate_data = false)
            end
        end
    end
end
```

**Comment by francispoulin** [comment]
@simone-silvestri : did you get a change to try deleting the files and running the test again?

**Comment by francispoulin** [comment]
@simone-silvestri and @glwagner, does someone want to try and removing the data and retrying the test?

**Comment by glwagner** [comment]
i think the data is installed freshly every time buildkite launches a new job, isn't it?

**Comment by glwagner** [comment]
Did you test on a GPU?

**Comment by glwagner** [comment]
Maybe this is resolved?

**Comment by navidcy** [comment]
They haven't failed for a while... I'm closing this for now.

**Comment by francispoulin** [comment]
Great to hear the tests are working!

**Comment by navidcy** [comment]
@francispoulin did you push any `.jld2`/`.nc` files on a branch when we were working on this?

**Comment by navidcy** [comment]
> They haven't failed for a while... I'm closing this for now.

they were STILL commented out! I'm opening and will close if indeed these tests don't constitute issue any more!

**Comment by navidcy** [comment]
Should we close after #3050?

**Comment by glwagner** [comment]
For sure because there are no regression tests
